README.md
# andOTP-decrypt

A backup decryptor for the [andOTP](https://github.com/andOTP/andOTP) Android app.

The tools in this package support the password based backup files of andOTP in both the current (0.6.3) old (0.6.2 and before) format.

Tools:

- `andotp_decrypt.py`: A decryption tool for password-secured backups of the [andOTP](https://github.com/flocke/andOTP) two-factor android app.
  - Output is written to stdout
- `generate_qr_codes.py`: A tool to generate new, scanable QR code images for every entry of a dump
  - Images are saved to the current working directory
- `generate_code.py`: A tool to generate a TOTP token for an account in the backup

## Installation

`pip install andotp-decrypt`

The tools will be installed as:

- `andotp_decrypt`
- `andotp_gencode`
- `andotp_qrcode`

## Development Setup

[Poetry](https://python-poetry.org/) install (recommended)

- Install poetry
  - `pip install poetry` (or use the recommended way from the website)
- Install everything else
  - `poetry install`
- Launch the virtualenv
  - `poetry shell`

Pip install

- `sudo pip3 install -r requirements.txt` 

On debian/ubuntu this should work:

- `sudo apt-get install python3-pycryptodome python3-pyotp python3-pyqrcode python3-pillow python3-docopt`

## Usage

- Dump JSON to the console:
  - `./andotp_decrypt.py /path/to/otp_accounts.json.aes`
- Generate new QR codes:
  - `./generate_qr_codes.py /path/to/otp_accounts.json.aes`
- Generate a TOTP code for your google account:
  - `./generate_code.py /path/to/otp_accounts.json.aes google`

## Thanks

Thank you for contributing!

- @alkuzad
- @ant9000
- @anthonycicc
- @erik-h
- @romed
- @rubenvdham
- @wornt
- @naums
- @marcopaganini


CHANGELOG.md
# Changelog

## 0.1.6
### Fixed
- Issuer matching for code generation

## 0.1.5
### Added
- Handle `issuer` field in JSON data (thanks to @naum and @marcopaganini)
- Add proper python tests (`testdata/test.sh` is still there as some kind of integration test)

## 0.1.4
### Added
- Emit better errors when run on empty files (thanks @wornt)
- poetry.lock added to git


testdata/test.sh
#!/bin/bash
set -e

D="$( cd "$(dirname "$0")" >/dev/null 2>&1 ; pwd -P )"
DATA_DIR=$(mktemp -d)

trap "echo 'Removing data dir'; rm -r $DATA_DIR;" exit

echo "Using directory $DATA_DIR"

cd $DATA_DIR

echo 123456 | python3 $D/../andotp_decrypt.py $D/accounts_new_123456.json.aes
echo 123456 | python3 $D/../generate_code.py $D/accounts_new_123456.json.aes test
echo 123456 | python3 $D/../generate_qr_codes.py $D/accounts_new_123456.json.aes

rm *.svg
echo 123456 | python3 $D/../andotp_decrypt.py -o $D/accounts_old_123456.json.aes
echo 123456 | python3 $D/../generate_code.py -o $D/accounts_old_123456.json.aes test
echo 123456 | python3 $D/../generate_qr_codes.py -o $D/accounts_old_123456.json.aes

rm *.svg
echo 123456 | python3 $D/../andotp_decrypt.py $D/accounts_issuer_123456.json.aes
echo 123456 | python3 $D/../generate_code.py $D/accounts_issuer_123456.json.aes test
echo 123456 | python3 $D/../generate_qr_codes.py $D/accounts_issuer_123456.json.aes
svgexpect=7
svgcount=$(ls *.svg | wc -l)
if [[ ! $svgcount -eq $svgexpect ]]; then
  echo "Error generating individual images, $svgcount images found, expected $svgexpect"
  ls -l *.svg
  exit 1
fi

rm *.svg
# This does not fail as it does not decode the data
echo 123456 | python3 $D/../andotp_decrypt.py $D/emptyfile.json.aes

# The following must all fail
if echo "123456" | python3 $D/../generate_code.py $D/emptyfile.json.aes test ; then exit 1; fi
if echo "123456" | python3 $D/../generate_qr_codes.py $D/emptyfile.json.aes ; then exit 1; fi

echo "123456" | python3 $D/../andotp_decrypt.py $D/garbage.json.aes
if echo "123456" | python3 $D/../generate_code.py $D/garbage.json.aes test ; then exit 1; fi
if echo "123456" | python3 $D/../generate_qr_codes.py $D/garbage.json.aes ; then exit 1; fi


tests/test_andotp.py
import pytest
import andotp_decrypt as ad
import json
from functools import partial

file_password = b'123456'


@pytest.fixture
def json_with_issuer():
    data = ad.decrypt_aes_new_format(file_password, 'testdata/accounts_issuer_123456.json.aes')
    yield json.loads(data)


@pytest.fixture(params=[True, False])
def decryptor(request):
    if request.param:
        yield partial(ad.decrypt_aes, input_file='testdata/accounts_old_123456.json.aes')
    else:
        yield partial(ad.decrypt_aes_new_format, input_file='testdata/accounts_new_123456.json.aes')


@pytest.fixture
def decrypted_json(decryptor):
    yield json.loads(decryptor(file_password))


def test_decrypt(decryptor):
    # The payload in the file should be a JSON doc of 160 characters
    assert len(decryptor(file_password)) == 160


def test_wrong_password(decryptor, capsys):
    assert not decryptor(b'wrong password')
    assert capsys.readouterr().out == "MAC check failed\nThe passphrase was probably wrong\n"


def test_empty_file(capsys):
    test_file = 'testdata/emptyfile.json.aes'
    assert not ad.decrypt_aes_new_format(file_password, test_file)
    assert capsys.readouterr().out == "No data could be read. The input file is unreadable or empty\n"


def test_json(decryptor):
    data = json.loads(decryptor(file_password))
    assert len(data) == 1
    assert data[0]['label'] == 'test'


def test_missing_issuer(decrypted_json):
    """Issuer fields were not always included as separate
       fields in the JSON data. E.g. the sample backups
       for old and new encryption do not contain the field."""
    assert 'issuer' not in decrypted_json[0]


def test_existing_issuer_field(json_with_issuer):
    assert len(json_with_issuer) == 7
    for entry in json_with_issuer:
        assert 'issuer' in entry


def test_find_entries(json_with_issuer):
    asmw_matches = ad.find_entries(json_with_issuer, 'asmw.org')
    assert len(asmw_matches) == 3

    ACME_matches = ad.find_entries(json_with_issuer, 'acme')
    assert len(ACME_matches) == 2

    # Matches are case insensitive
    ACME_matches = ad.find_entries(json_with_issuer, 'ACME')
    assert len(ACME_matches) == 2


def test_find_by_tag(json_with_issuer):
    tag_matches = ad.find_entries(json_with_issuer, 'sometag')
    assert len(tag_matches) == 2


def test_find_limit(json_with_issuer):
    tag_matches = ad.find_entries(json_with_issuer, 'asmw', 1)
    assert len(tag_matches) == 1

    tag_matches = ad.find_entries(json_with_issuer, 'asmw', 2)
    assert len(tag_matches) == 2


generate_code.py
#!/usr/bin/env python3
"""generate_code.py

Usage:
  generate_code.py [-o|--old] [-a] ANDOTP_AES_BACKUP_FILE MATCH_STRING

Options:
  -o --old      Use old encryption (andOTP <= 0.6.2)
  -a --all      Show all matches.
  -h --help     Show this screen.
  --version     Show version.

"""
from docopt import docopt
import sys
import pyotp
import json
import andotp_decrypt


def main():
    arguments = docopt(__doc__, version='generate_code 0.1')

    password = andotp_decrypt.get_password()
    text = None
    if arguments['--old']:
        text = andotp_decrypt.decrypt_aes(password, arguments['ANDOTP_AES_BACKUP_FILE'])
    else:
        text = andotp_decrypt.decrypt_aes_new_format(password, arguments['ANDOTP_AES_BACKUP_FILE'])

    if not text:
        print("Something went wrong while loading %s. Maybe the passphrase was wrong"
              " or the input file is empty." % arguments['ANDOTP_AES_BACKUP_FILE'])
        sys.exit(1)
    entries = json.loads(text)

    filtered_entries = andotp_decrypt.find_entries(entries, arguments["MATCH_STRING"], None if arguments["--all"] else 1)
    if len(filtered_entries) == 0:
        print("No entry matching '%s' found" % arguments["MATCH_STRING"])
        sys.exit(0)

    for entry in filtered_entries:
        if entry['type'] == 'TOTP':
            totp = pyotp.TOTP(entry['secret'], interval=entry['period'])
            print("Matched: %s" % andotp_decrypt.descriptor(entry))
            print(totp.now())
        else:
            print("Unsupported OTP type: %s" % entry["type"])
            sys.exit(2)


if __name__ == '__main__':
    main()


andotp-decrypt.py
#!/usr/bin/env python3
from aes_gcm.aes_gcm import AES_GCM, InvalidTagException
from Crypto.Hash import SHA256
from os.path import basename
from getpass import getpass
import sys
import os

debug = False

def bytes2Hex(bytes):
    return '(%s) 0x%s' % (len(bytes), ''.join('{:02x}'.format(x) for x in bytes))

def usage():
    print("Usage: %s /path/to/otp_accounts.json.aes" % basename(sys.argv[0]))
    os._exit(0)

if len(sys.argv) != 2:
    usage()

input_file = sys.argv[1]
if not os.path.exists(input_file):
    print("Could not find input file: %s" % input_file)
    usage()

pw = getpass('andOTP passphrase:')
hash = SHA256.new(pw.strip().encode('UTF-8'))
symmetric_key = hash.digest()
if debug:
    print("Symmetric key: %s" % bytes2Hex(symmetric_key))

input_bytes = None
with open(input_file,'rb') as f:
    input_bytes = f.read()

# Raw data structure is IV[:12] + crypttext[12:-16] + auth_tag[-16:]
iv = input_bytes[:12]
crypttext = input_bytes[12:-16]
tag = input_bytes[-16:]
if debug:
    print("Input bytes: %", bytes2Hex(input_bytes))
    print("IV: %s" % bytes2Hex(iv))
    print("Crypttext: %s" % bytes2Hex(crypttext))
    print("Auth tag: %s" % bytes2Hex(tag))

aes = AES_GCM(symmetric_key)
try:
    dec = aes.decrypt(iv, crypttext, tag)
except InvalidTagException as e:
    print(e)
    print("The passphrase was probably wrong")
    os._exit(1)
if debug:
    print("Decrypted data: %s" % bytes2Hex(dec))

print(dec.decode('UTF-8'))


generate_qr_codes.py
#!/usr/bin/env python3
"""generate_qr_codes.py

Usage:
  generate_qr_codes.py [-o|--old] ANDOTP_AES_BACKUP_FILE

Options:
  -o --old      Use old encryption (andOTP <= 0.6.2)
  -h --help     Show this screen.
  --version     Show version.

"""
from docopt import docopt
import sys
import pyotp
import pyqrcode
import json
import andotp_decrypt
import os.path


def valid_filename_char(c):
    return c.isalpha() or c.isdigit() or c in "@_-"


def gen_filename(entry):
    parts = []
    label = entry.get('label')
    if label:
        parts.append(label)
    issuer = entry.get('issuer')
    if issuer:
        parts.append(issuer)

    key = '@'.join(parts)

    prefix = "".join([c for c in key if valid_filename_char(c)]).strip()

    candidate = f"{prefix}.svg"
    counter = 0
    while os.path.exists(candidate):
        counter += 1
        candidate = f"{prefix}_{counter}.svg"

    return candidate


def main():
    arguments = docopt(__doc__, version='generate_qr_codes 0.1')

    password = andotp_decrypt.get_password()
    text = None
    if arguments['--old']:
        text = andotp_decrypt.decrypt_aes(password, arguments['ANDOTP_AES_BACKUP_FILE'])
    else:
        text = andotp_decrypt.decrypt_aes_new_format(password, arguments['ANDOTP_AES_BACKUP_FILE'])

    if not text:
        print("Something went wrong while loading %s. Maybe the passphrase was wrong?"
              " or the input file is empty!" % arguments['ANDOTP_AES_BACKUP_FILE'])
        sys.exit(1)
    entries = json.loads(text)
    for entry in entries:
        url = None
        issuer = entry.get('issuer')
        label = entry.get('label')
        if not issuer and " - " in label:
            issuer, label = label.split(" - ", 1)
        if entry['type'] == 'TOTP':
            totp = pyotp.TOTP(entry['secret'], interval=entry['period'])
            url = totp.provisioning_uri(label, issuer_name=issuer)
        elif entry['type'] == 'HOTP':
            totp = pyotp.TOTP(entry['secret'])
            url = totp.provisioning_uri(label, issuer_name=issuer)
        if url:
            img = pyqrcode.create(url)
            save_filename = gen_filename(entry)
            img.svg(save_filename, scale=4, background='#fff')
            print("Code saved as: %s" % save_filename)


if __name__ == '__main__':
    main()


andotp_decrypt.py
#!/usr/bin/env python3
"""andotp-decrypt.py

Usage:
andotp-decrypt.py [-o|--old] [--debug] [-h|--help] [--version] INPUT_FILE

Options:
-o --old      Use old encryption (andOTP <= 0.6.2)
--debug       Print debug info
-h --help     Show this screen.
--version     Show version.

"""

import os
import sys
import hashlib
import struct
from getpass import getpass

from Crypto.Cipher import AES
from Crypto.Hash import SHA256

from docopt import docopt


def bytes2Hex(bytes2encode):
    return '(%s) 0x%s' % (len(bytes2encode), ''.join('{:02x}'.format(x) for x in bytes2encode))


def decode(key, data, debug=False):
    """Decode function used for both the old and new style encryption"""
    # Raw data structure is IV[:12] + crypttext[12:-16] + auth_tag[-16:]
    iv = data[:12]
    crypttext = data[12:-16]
    tag = data[-16:]
    if debug:
        print("Input bytes: %", bytes2Hex(data))
        print("IV: %s" % bytes2Hex(iv))
        print("Crypttext: %s" % bytes2Hex(crypttext))
        print("Auth tag: %s" % bytes2Hex(tag))
    try:
        aes = AES.new(key, AES.MODE_GCM, nonce=iv)
    except Exception as e:
        print(e)
        return None

    try:
        dec = aes.decrypt_and_verify(crypttext, tag)
        if debug:
            print("Decrypted data: %s" % bytes2Hex(dec))
        return dec.decode('UTF-8')
    except ValueError as e:
        print(e)
        print("The passphrase was probably wrong")
        return None


def decrypt_aes_new_format(password, input_file, debug=False):
    input_bytes = None
    with open(input_file, 'rb') as f:
        input_bytes = f.read()

    if len(input_bytes) == 0:
        print("No data could be read. The input file is unreadable or empty")
        return None

    try:
        # Raw data structure is iterations[:4] + salt[4:16] + data[16:]
        iterations = struct.unpack(">I", input_bytes[:4])[0]
        salt = input_bytes[4:16]
        data = input_bytes[16:]
        try:
            pbkdf2_key = hashlib.pbkdf2_hmac('sha1', password, salt, iterations, 32)
        except OverflowError:
            print("Could not parse iterations from file, "
                  "either the file is broken or the format has changed.")
            return None
        if debug:
            print("Iterations: %s" % iterations)
            print("Salt: %s" % bytes2Hex(salt))
            print("Pbkdf2 key: %s" % bytes2Hex(pbkdf2_key))
        return decode(pbkdf2_key, data, debug)
    except struct.error:
        print("The input data could not be decrypted")
        return None


def decrypt_aes(password, input_file, debug=False):
    hash = SHA256.new(password)
    symmetric_key = hash.digest()
    if debug:
        print("Symmetric key: %s" % bytes2Hex(symmetric_key))
    input_bytes = None
    with open(input_file, 'rb') as f:
        input_bytes = f.read()

    return decode(symmetric_key, input_bytes, debug)


def get_password():
    # Read the backup password manually via a prompt if stdin is a tty,
    # otherwise read from stdin directly.
    if sys.stdin.isatty():
        pw = getpass('andOTP AES passphrase:')
    else:
        pw = sys.stdin.readline()
    return pw.strip().encode('UTF-8')


def find_entries(data, pattern, limit=None):
    result = []

    for entry in data:
        label = entry['label']
        # NOTE: issuer was not always part of the JSON structure
        issuer = entry.get('issuer', '')
        tags = "".join(entry.get('tags', []))
        key = label + issuer + tags
        if pattern.lower() in key.lower():
            result.append(entry)
            if limit and len(result) == limit:
                break

    return result


def descriptor(entry):
    label = entry.get('label')
    issuer = entry.get('issuer')
    if label and issuer:
        return f'{label}[{issuer}]'
    elif label:
        return label
    elif issuer:
        return f'[{issuer}]'
    else:
        return 'no label or issuer found'


def main():
    arguments = docopt(__doc__, version='andotp-decrypt 0.1')
    input_file = arguments['INPUT_FILE']
    debug = arguments['--debug']
    old_encryption = arguments['--old']
    if not os.path.exists(input_file):
        print("Could not find input file: %s" % input_file)
        return None
    password = get_password()
    if old_encryption:
        print(decrypt_aes(password, input_file, debug))
    else:
        print(decrypt_aes_new_format(password, input_file, debug))


if __name__ == '__main__':
    main()
